### ISSUE TITLE
Bug: [consistent-return] function with exhaustive switch case returns reports a false positive

### ISSUE THREAD

**Issue by sharunkumar** [typescript-eslint/typescript-eslint#9708]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.ts&code=MYewdgzgLgBAtgTwIICcUEMEwLwwNoDkAZiCAQDQwEBG6KFVtAXgQLozoQyiRQBQfAPSCYUABYBLLgFM0IFFxABXfj2jwEAMSVhgUCeBwwAFBKjS4ALlEIADtJBENqDAjxglcarNYBKa9AoEmAA5jgAfDAA3nwwMBAA7mbAYiZmFr7RsXHcnNJUJGTWKNJQSihgVAA28hYEANzZccB5jHQExaXllQQSthCeDU25EPk06CydZRXVtXBDcQC%2BfMtCIuJSMGKcMDol6Cno1FX5oAAm0nxqsIjauvrgAKpg%2B4fH%2Bbim5lY29o7OaEw7k83hQfgCUCCoQiWTiiWSqS%2BGVhORaowKpA6MBK0x6NRK80aORGY1o9Cm3SofQGhOGaNJEyxOMpBHxdSJMGWcWZMwIDRgwhgYBAsFe23eKyAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQMYHsA7ZAS2QBdFDyBaaRc2aQ9KfAM3cgBpxtIAAuQCeABxS5oJUbRTwS1APQFiZStToMmLDJETRo%2BaJD4BfEKaA&tsconfig=&tokens=false

### Repro Code

```TypeScript
const myArray = ['foo', 'bar', 'baz'] as const

// this errors out
const myFunction = (item: typeof myArray[number]): string => {
  switch (item) {
    case 'foo': return 'lorem';
    case 'bar': return 'ipsum';
    case 'baz': return 'lorem';
  }
}

// this has unreachable code
const myFunctionUnreachable = (item: typeof myArray[number]): string => {
  switch (item) {
    case 'foo': return 'lorem';
    case 'bar': return 'ipsum';
    case 'baz': return 'lorem';
  }
  return ''; // not reachable
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "consistent-return": "off",
    "@typescript-eslint/consistent-return": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

In this case, putting a return value at the end of the function would lead to unreachable code

### Actual Result

linter would not report an error

**Comment by bradzacher** [comment]
Please see the comment at the top of the rule docs
https://typescript-eslint.io/rules/consistent-return/

This is expected and the rule is not recommended to be used. This is not something that will be fixed.
